scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -338,12 +369,123 @@ def highly_variable_genes(
             ('dispersions_norm', 'float32'),
         ]
         if batch_key is not None:
-            arrays.extend([
-                df['highly_variable_nbatches'].values,
-                df['highly_variable_intersection'].values,
-            ])
-            dtypes.append([
-                ('highly_variable_nbatches', int),
-                ('highly_variable_intersection', np.bool_),
-            ])
+            arrays.extend(
+                [
+                    df['highly_variable_nbatches'].values,
+                    df['highly_variable_intersection'].values,
+                ]
+            )
+            dtypes.append(
+                [
+                    ('highly_variable_nbatches', int),
+                    ('highly_variable_intersection', np.bool_),
+                ]
+            )
         return np.rec.fromarrays(arrays, dtype=dtypes)
+
+
+def highly_variable_genes_seurat_v3(
+    adata: AnnData,
+    n_top_genes: int = 2000,
+    batch_key: Optional[str] = None,
+    lowess_frac: Optional[float] = 0.15,
+):
+    """\
+    Annotate highly variable genes [Stuart19]_.
+
+    Expects raw count data.
+
+    The major difference in this implementation is the use of lowess insted of loess.
+
+    For further details of the sparse arithmetic see https://www.overleaf.com/read/ckptrbgzzzpg
+    Parameters
+    ----------
+    adata
+        The annotated data matrix of shape `n_obs` × `n_vars`. Rows correspond
+        to cells and columns to genes.
+    n_top_genes
+        Number of highly-variable genes to keep.
+    batch_key
+        If specified, highly-variable genes are selected within each batch separately and merged.
+        This simple process avoids the selection of batch-specific genes and acts as a
+        lightweight batch correction method.
+    lowess_frac
+        The fraction of the data (cells) used when estimating the variance in the lowess model fit.
+    """
+    import statsmodels
+
+    lowess = statsmodels.nonparametric.lowess
+
+    if batch_key is None:
+        batch_info = pd.Categorical(np.zeros((adata.X.shape[0])).astype(int))
+    else:
+        batch_info = adata.obs[batch_key]
+
+    norm_gene_vars = []
+    for b in np.unique(batch_info):
+
+        mean, var = materialize_as_ndarray(_get_mean_var(adata[batch_info == b].X))
```

Comment:
What does the `materialize_as_ndarray` do here?

```suggestion
        mean, var = _get_mean_var(adata[batch_info == b].X)
```

Reply:
This I'm not sure about, I see it in the current hvg function, as well as a call to `sanitize_adata`. Is it necessary?

Reply:
`sanitize_anndata` should change views of anndata objects into actual ones, and will clean up categorical values.

`materialize_as_ndarray` is mostly used for forcing distributed or lazy arrays into memory. It's only necessary if this function would work for dask arrays.